### PR TITLE
Tweak: Load nested elements styles only when they are in use [ED-15011]

### PR DESCRIPTION
**Pull request by louiswol94**

### CODE REVIEW

File: .grunt-config/modules-with-widgets-css.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
+const path = require( 'path' );
```
Comment (louiswol94) [inline .grunt-config/modules-with-widgets-css.js:1]: Copied this file from Pro to handle SCSS in `modules/**/assets/scss/frontend`
The current grunt tasks in Core only catered for `assets/dev/scss/frontend/widgets`

Now the fact that these css files will be compiled in `elementor/assets/css/widget-nested-tabs.min.css`, the Improved Loading Experiment will make sure it is printed inside the widget div.

So the issue was just that modules SCSS files were never compiled.

File: assets/dev/scss/frontend/_modules.scss [hunk 2]
```diff
@@ -1,6 +1,4 @@
 // Shapes Module
 @import '../../../../modules/shapes/assets/scss/frontend';
-@import '../../../../modules/nested-tabs/assets/scss/frontend';
```
Comment (louiswol94) [inline assets/dev/scss/frontend/_modules.scss:3]: This is now be removed out of the `frontend-lite` file.

File: .grunt-config/modules-with-widgets-css.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		} );
+	}
+
+	getWidgetScssContent( widgetName, direction, widgetPath ) {
+		return `$direction: ${ direction };
+
+@import "../../../../../elementor/assets/dev/scss/helpers/direction";
+@import "../../../../../elementor/assets/dev/scss/helpers/variables";
+@import "../../../../../elementor/assets/dev/scss/frontend/breakpoints/breakpoints";
+@import "../../../../../elementor/assets/dev/scss/helpers/mixins";
```
Comment (louiswol94) [inline .grunt-config/modules-with-widgets-css.js:75]: @hein-obox With the we won't need to import this in each file individually :)

File: modules/nested-tabs/widgets/nested-tabs.php [hunk 4]
```diff
@@ -79,10 +79,10 @@ protected function register_controls() {
 		$end = is_rtl() ? 'left' : 'right';
 		$start_logical = is_rtl() ? 'end' : 'start';
 		$end_logical = is_rtl() ? 'start' : 'end';
-		$heading_selector_non_touch_device = '{{WRAPPER}} > .elementor-widget-container > .e-n-tabs[data-touch-mode="false"] > .e-n-tabs-heading';
-		$heading_selector_touch_device = '{{WRAPPER}} > .elementor-widget-container > .e-n-tabs[data-touch-mode="true"] > .e-n-tabs-heading';
-		$heading_selector = '{{WRAPPER}} > .elementor-widget-container > .e-n-tabs > .e-n-tabs-heading';
-		$content_selector = ':where( {{WRAPPER}} > .elementor-widget-container > .e-n-tabs > .e-n-tabs-content ) > .e-con';
+		$heading_selector_non_touch_device = '{{WRAPPER}}.elementor-widget-n-tabs > .elementor-widget-container > .e-n-tabs[data-touch-mode="false"] > .e-n-tabs-heading';
```
Comment (louiswol94) [inline modules/nested-tabs/widgets/nested-tabs.php:82]: We need to make the control selectors more specific, so that the widget CSS does not override it.

File: assets/dev/scss/frontend/_widgets.scss [hunk 5]
```diff
@@ -24,3 +24,9 @@
 @import "widgets/toggle";
 @import "widgets/video";
 @import "widgets/rating";
+
+//
+// Modules with Widgets
+//
+@import '../../../../modules/nested-tabs/assets/scss/frontend';
+@import '../../../../modules/nested-accordion/assets/scss/frontend';
```
Comment (rami-elementor) [inline assets/dev/scss/frontend/_widgets.scss:32]: @louiswol94 Maybe it should be placed at the top of the file (line 1), above the widget, to keep the same loading order? For BC.

File: modules/nested-tabs/widgets/nested-tabs.php (hunk 4 above)
Comment (rami-elementor) [reply modules/nested-tabs/widgets/nested-tabs.php:82]: Only on nested tabs? what about nested accordion?

File: assets/dev/scss/frontend/_widgets.scss (hunk 5 above)
Comment (louiswol94) [reply assets/dev/scss/frontend/_widgets.scss]: @rami-elementor Sure, updated!

File: modules/nested-tabs/widgets/nested-tabs.php (hunk 4 above)
Comment (louiswol94) [reply modules/nested-tabs/widgets/nested-tabs.php:82]: @rami-elementor Yes, this is not like a general issue. The developer just added too specific selectors in the Nested tabs SCSS file.

**Review by hein-obox — DISMISSED** [review]

**Review by rami-elementor — DISMISSED** [review]

**Review by rotemee — CHANGES_REQUESTED** [review]

File: .grunt-config/modules-with-widgets-css.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
+const path = require( 'path' );
+const fs = require( 'fs' );
+
+class ModulesWithWidgetsCss {
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/modules-with-widgets-css.js:4]: Hey @rotemee, thanks for the review. I fully understand your reasoning, and after our discussion I tried to implement this, but in reality the logic is too different between the stand alone widgets and modules that it became to convoluted to combine it in one file. Literally every single method in the file had to be changed (so it cannot really be seen as a full duplication of the existing file in core) - I had to keep track of each type of widget and put a bunch of if/else statements due to the different structure. It became really hard to follow, and I could not get it to run fully yet without errors. It actually made more sense the way it is implemented now. Seeing that our sprint ends today, I would be happy if we can approve this PR to not block the epic, and continue our discussions offline. I will take ownership of this :)

[the remark this answers is not in the document]
Comment (rotemee) [reply .grunt-config/modules-with-widgets-css.js:4]: @louiswol94 - Let's talk about it :)

[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/modules-with-widgets-css.js]: @rotemee Updated to only used one file. Thanks for pushing.

**Review by hein-obox — DISMISSED** [review]

File: .grunt-config/widgets-css.js [hunk 7]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
 
+	_getStandAloneWidgetsCssFilesList() {
+		const standAloneWidgetData = [],
+			standAloneSidgetsList = fs.existsSync( this.sourceScssFolder ) ? fs.readdirSync( this.sourceScssFolder ) : [];
+
+		standAloneSidgetsList.forEach( ( filename ) => {
+			const widgetName = filename.replace( '.scss', '' );
+
+			standAloneWidgetData.push( {
+				widgetName: widgetName,
```
Comment (rotemee) [inline .grunt-config/widgets-css.js:107]:
```suggestion
				widgetName,
```

File: .grunt-config/widgets-css.js [hunk 8]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+				defaultFilename: this.cssFilePrefix + filename,
+				rtlFilename: this.cssFilePrefix + filename.replace( '.scss', '-rtl.scss' ),
```
Comment (rotemee) [inline .grunt-config/widgets-css.js:109]: You can now remove the `getCssFileNames` function for not being used anywhere.

File: .grunt-config/widgets-css.js [hunk 9]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+
+		return frontendScssFiles;
+	}
+
+	getWidgetDataFromPath( baseFolder, filePath ) {
+		// Removing base-folder and first slash so that the module name will be the first value in the path.
+		filePath = filePath.replace( baseFolder, '' ).substring(1);
+
+		return {
+			path: filePath.replace( /\\/g, '/' ),
```
Comment (rotemee) [inline .grunt-config/widgets-css.js:158]: Maybe I'm missing something but why do we need this replace?

**Review by rotemee — CHANGES_REQUESTED** [review]

File: .grunt-config/widgets-css.js (hunk 9 above)
Comment (louiswol94) [reply .grunt-config/widgets-css.js:149]: @rotemee Not sure. Copied this from Pro.

File: .grunt-config/widgets-css.js (hunk 8 above)
Comment (louiswol94) [reply .grunt-config/widgets-css.js:102]: Done

File: .grunt-config/widgets-css.js (hunk 7 above)
Comment (louiswol94) [reply .grunt-config/widgets-css.js]: Done

File: .grunt-config/widgets-css.js [hunk 10]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+		moduleWidgetsList.forEach( ( filePath ) => {
+			const widgetData = this.getWidgetDataFromPath( this.sourceModulesScssFolder, filePath );
+
+			moduleWidgetData.push( {
+				widgetName: widgetData.name,
+				defaultFilename: this.cssFilePrefix + widgetData.name + '.scss',
+				rtlFilename: this.cssFilePrefix + widgetData.name + '-rtl.scss',
+				sourceFolder: this.sourceModulesScssFolder,
+				importPath: `../../../../modules/${  widgetData.path }`,
+				filePath: filePath,
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/widgets-css.js]: Done

File: .grunt-config/widgets-css.js [hunk 11]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+			} );
+		} );
+
+		return moduleWidgetData;
+	}
+
+	getModulesFrontendScssFiles( filePath, frontendScssFiles = [] ) {
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/widgets-css.js:130]: Removed the prefixes rather

File: .grunt-config/widgets-css.js [hunk 12]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+			standAloneSidgetsList = fs.existsSync( this.sourceScssFolder ) ? fs.readdirSync( this.sourceScssFolder ) : [];
+
+		standAloneSidgetsList.forEach( ( filename ) => {
+			const widgetName = filename.replace( '.scss', '' );
+
+			standAloneWidgetData.push( {
+				widgetName: widgetName,
+				defaultFilename: this.cssFilePrefix + filename,
+				rtlFilename: this.cssFilePrefix + filename.replace( '.scss', '-rtl.scss' ),
+				sourceFolder: this.sourceScssFolder,
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/widgets-css.js]: Removed

File: .grunt-config/widgets-css.js [hunk 13]
```diff
@@ -92,11 +91,75 @@ class WidgetsCss {
...
+		this.widgetsScssFilesList = this._getStandAloneWidgetsCssFilesList().concat( this._getModulesWidgetsCssFilesList() );
 
 		return this.widgetsScssFilesList;
 	}
 
+	_getStandAloneWidgetsCssFilesList() {
+		const standAloneWidgetData = [],
+			standAloneSidgetsList = fs.existsSync( this.sourceScssFolder ) ? fs.readdirSync( this.sourceScssFolder ) : [];
+
+		standAloneSidgetsList.forEach( ( filename ) => {
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply .grunt-config/widgets-css.js]: Typo.
Fixed.

**Review by rotemee — APPROVED** [review]
